File: Android/RTMSyncManager/src/main/java/io/agora/rtmsyncmanager/service/arbiter/AUIArbiter.kt
package io.agora.rtmsyncmanager.service.arbiter

import io.agora.rtmsyncmanager.service.rtm.AUIRtmException
import io.agora.rtmsyncmanager.service.rtm.AUIRtmLockRespObserver
import io.agora.rtmsyncmanager.service.rtm.AUIRtmManager
import io.agora.rtmsyncmanager.utils.AUILogger
import io.agora.rtmsyncmanager.utils.ObservableHelper

/**
 * AUIArbiter is a class that manages the arbitration process in a channel.
 * It handles the creation, destruction, acquisition, and release of locks in a channel.
 * It also notifies the subscribed handlers about the changes in the arbitration process.
 */
class AUIArbiter(
    private val channelName: String,
    private val rtmManager: AUIRtmManager,
    private val currentUserId: String
) {

    private val tag = "AUIArbiter"

    private var lockOwnerId = ""
        set(value) {
            field = value
            notifyArbiterDidChange()
        }
    private var arbiterHandlers = ObservableHelper<AUIArbiterCallback>()

    private val rtmLockRespObserver = object : AUIRtmLockRespObserver {
        override fun onReceiveLock(channelName: String, lockName: String, lockOwner: String) {
            AUILogger.logger().d(tag, "onReceiveLock channelName: $channelName lockName: $lockName lockOwner: $lockOwner")
            if (channelName != this@AUIArbiter.channelName) return
            /*
             * The following two scenarios require refreshing the metadata to the latest:
             * 1. If the lockOwnerId is the current user and the arbiter has switched (not the first acquisition, otherwise the first onReceiveLock obtained in roomService is empty),
             *    it is necessary to refresh before notifying the external lock transfer.
             * 2. If the lockOwnerId is not the current user, and the previous lockOwnerId was the current user, it means that the current user has switched from the arbiter to a non-arbiter,
             *    and it needs to be refreshed after notifying the external (because the collection will not use remote data when it thinks it is the lock owner),
             *    it is possible that the local data of the current user is not up to date.
             */
            val gotLockFromOthers = lockOwner == currentUserId
            val lossLockToOthers = lockOwnerId == currentUserId && lockOwner != currentUserId
            if (gotLockFromOthers) {
                rtmManager.fetchMetaDataSnapshot(channelName) {
                    //TODO: error handler, retry?
                    lockOwnerId = lockOwner
                }
            } else if (lossLockToOthers) {
                lockOwnerId = lockOwner
                rtmManager.fetchMetaDataSnapshot(channelName) {
                    //TODO: error handler, retry?
                }
            } else {
                lockOwnerId = lockOwner
            }
        }

        override fun onReleaseLock(channelName: String, lockName: String, lockOwner: String, isExpire: Boolean) {
            AUILogger.logger().d(tag, "onReleaseLock channelName: $channelName lockName: $lockName lockOwner: $lockOwner")
            if (channelName == this@AUIArbiter.channelName) {
                acquire()
                if (isExpire && lockOwnerId != currentUserId) return
                lockOwnerId = ""
            }
        }
    }

    init {
        rtmManager.subscribeLock(channelName, observer = rtmLockRespObserver)
    }

    fun deInit(){
        rtmManager.unsubscribeLock(rtmLockRespObserver)
    }

    // Subscribes a handler to the arbiter events
    fun subscribeEvent(handler: AUIArbiterCallback) {
        arbiterHandlers.subscribeEvent(handler)
    }

    // Unsubscribes a handler from the arbiter events
    fun unSubscribeEvent(handler: AUIArbiterCallback) {
        arbiterHandlers.unSubscribeEvent(handler)
    }

    // Creates a lock in the channel
    fun create(completion: ((AUIRtmException?) -> Unit)? = null) {
        AUILogger.logger().d(tag, "setLock start: $channelName")
        rtmManager.setLock(channelName) { error ->
            AUILogger.logger().d(tag, "setLock error: $error")
            notifyError(error)
            completion?.invoke(error)
        }
    }

    // Destroys the lock in the channel
    fun destroy(completion: ((AUIRtmException?) -> Unit)? = null) {
        AUILogger.logger().d(tag, "removeLock channelName: $channelName")
        rtmManager.removeLock(channelName) { error ->
            AUILogger.logger().d(tag, "removeLock error: $error")
            notifyError(error)
            completion?.invoke(error)
        }
    }

    // Acquires the lock in the channel
    fun acquire(completion: ((AUIRtmException?) -> Unit)? = null) {
        AUILogger.logger().d(tag, "acquireLock channelName: $channelName")
        rtmManager.acquireLock(channelName) { error ->
            AUILogger.logger().d(tag, "acquireLock error: $error")
            notifyError(error)
            completion?.invoke(error)
        }
    }

    // Releases the lock in the channel
    fun release(completion: ((AUIRtmException?) -> Unit)? = null) {
        AUILogger.logger().d(tag, "releaseLock channelName: $channelName")
        rtmManager.releaseLock(channelName){ error ->
            AUILogger.logger().d(tag, "releaseLock error: $error")
            notifyError(error)
            completion?.invoke(error)
        }
    }

    // Checks if the current user is the arbiter
    fun isArbiter() = lockOwnerId == currentUserId

    // Returns the ID of the user who currently owns the lock
    fun lockOwnerId() = lockOwnerId

    // Notifies the handlers about an error
    private fun notifyError(error: AUIRtmException?) {
        if (error == null) {
            return
        }
        arbiterHandlers.notifyEventHandlers { handler ->
            handler.onError(channelName, error)
        }
    }

    // Notifies the handlers when the arbiter changes
    private fun notifyArbiterDidChange() {
        arbiterHandlers.notifyEventHandlers { handler ->
            handler.onArbiterDidChange(channelName, lockOwnerId)
        }
    }
}

File: iOS/AgoraEntScenarios/Common/API/RTMSyncManager/RTMSyncManager/Classes/Arbiter/AUIArbiter.swift
//
//  AUIArbiter.swift
//  AUIKitCore
//
//  Created by wushengtao on 2023/11/2.
//

import AgoraRtmKit

@objc public class AUIArbiter: NSObject {
    private var channelName: String!
    private var rtmManager: AUIRtmManager!
    private var currentUserInfo: AUIUserThumbnailInfo!
    private(set) var lockOwnerId: String = "" {
        didSet {
            notifyArbiterDidChange()
        }
    }
    private var arbiterDelegates: NSHashTable<AUIArbiterDelegate> = NSHashTable<AUIArbiterDelegate>()
    
    deinit {
        aui_info("deinit AUIArbiter", tag: "AUIArbiter")
        rtmManager.unsubscribeLock(channelName: channelName, lockName: kRTM_Referee_LockName, delegate: self)
    }
    
    public init(channelName: String, rtmManager: AUIRtmManager, userInfo: AUIUserThumbnailInfo) {
        self.rtmManager = rtmManager
        self.channelName = channelName
        self.currentUserInfo = userInfo
        super.init()
        rtmManager.subscribeLock(channelName: channelName, lockName: kRTM_Referee_LockName, delegate: self)
        aui_info("init AUIArbiter", tag: "AUIArbiter")
    }
    
    public func subscribeEvent(delegate: AUIArbiterDelegate) {
        if arbiterDelegates.contains(delegate) { return }
        
        arbiterDelegates.add(delegate)
    }
    
    public func unSubscribeEvent(delegate: AUIArbiterDelegate) {
        arbiterDelegates.remove(delegate)
    }
    
    /// 创建锁
    public func create(completion: ((NSError?)-> ())? = nil) {
        rtmManager.setLock(channelName: channelName, lockName: kRTM_Referee_LockName) {[weak self] err in
            self?.notifyError(error: err)
            completion?(err)
        }
    }
    
    /// 销毁锁
    public func destroy(completion: ((NSError?)-> ())? = nil) {
        rtmManager.removeLock(channelName: channelName, lockName: kRTM_Referee_LockName) {[weak self] err in
            self?.notifyError(error: err)
            completion?(err)
        }
    }
    
    /// 获取锁
    public func acquire(completion: ((NSError?)-> ())? = nil) {
        rtmManager.acquireLock(channelName: channelName, lockName: kRTM_Referee_LockName) {[weak self] err in
            self?.notifyError(error: err)
            completion?(err)
        }
    }
    
    /// 释放锁
    public func release(completion: ((NSError?)-> ())? = nil) {
        rtmManager.releaseLock(channelName: channelName, lockName: kRTM_Referee_LockName) {[weak self] err in
            self?.notifyError(error: err)
            completion?(err)
        }
    }
    
    public func isArbiter() -> Bool {
        return lockOwnerId == currentUserInfo.userId
    }
}

//MARK: private
extension AUIArbiter {
    private func notifyError(error: NSError?) {
        guard let error = error else { return }
        aui_info("notifyError: \(error.localizedDescription)", tag: "AUIArbiter")
        arbiterDelegates.allObjects.forEach { delegate in
            delegate.onError(channelName: channelName, error: error)
        }
    }
    
    private func notifyArbiterDidChange() {
        arbiterDelegates.allObjects.forEach { delegate in
            delegate.onArbiterDidChange(channelName: channelName, arbiterId: self.lockOwnerId)
        }
    }
}

//MARK: AUIRtmLockProxyDelegate
extension AUIArbiter: AUIRtmLockProxyDelegate {
    public func onReceiveLockDetail(channelName: String, lockDetail: AgoraRtmLockDetail, eventType: AgoraRtmLockEventType) {
        aui_info("onReceiveLockDetail[\(channelName)]: \(lockDetail.owner)/\(currentUserInfo.userId)")
        guard channelName == self.channelName else {return}
        /*
         下列两种情况需要刷新下metadata到最新
         1. 如果lockOwnerId是自己，需要在通知外部锁转移前刷新下
         2. 如果lockOwnerId不是自己，而之前lockOwnerId是自己，说明自己从仲裁者切换成非仲裁者了，需要通知外部后刷新下(因为collection认为是锁主的情况下是不会用远端数据的)，可能自己的本地数据没有到最新
         */
        let gotLock = lockDetail.owner == currentUserInfo.userId
        let lossLockToOthers = lockOwnerId == currentUserInfo.userId && lockDetail.owner != currentUserInfo.userId
        if gotLock {
            rtmManager.fetchMetaDataSnapshot(channelName: channelName) {[weak self] error in
                guard let self = self else { return }
                //TODO: error handler, retry?
                self.lockOwnerId = lockDetail.owner
            }
        } else if lossLockToOthers {
            self.lockOwnerId = lockDetail.owner
            rtmManager.fetchMetaDataSnapshot(channelName: channelName) {[weak self] error in
                guard let self = self else { return }
                //TODO: error handler, retry?
                
            }
        } else {
            lockOwnerId = lockDetail.owner
        }
    }
    
    public func onReleaseLockDetail(channelName: String, lockDetail: AgoraRtmLockDetail, eventType: AgoraRtmLockEventType) {
        aui_info("onReleaseLockDetail[\(channelName)]: \(lockDetail.owner)")
        guard channelName == self.channelName else {return}
        rtmManager.acquireLock(channelName: channelName, lockName: kRTM_Referee_LockName) { err in
        }
        //过期可能会在获取锁之后收到，导致把正确的锁主清理了，因此只在锁主是自己的时候才处理
        if eventType == .lockExpired, lockOwnerId == currentUserInfo.userId {
            self.lockOwnerId = ""
        }
        
    }
}
